Proposed copy revisions (#2)
researchapps/betterscientificsoftware.github.io: Articles/Blog/2023-04-developer-surveys.md

#### Publication date: xxxxxxxxxxx

### Why run a developer survey?

If you've looked around the internet in the last decade, you've probably noticed that commercial organizations providing software development tooling tend to run surveys.
Some notable examples include the [Stack Overflow Developer Survey](https://survey.stackoverflow.co/), the [GitHub Octoverse Survey](https://octoverse.github.com/), or the [JetBrains State of the Developer Ecosystem](https://www.jetbrains.com/lp/devecosystem-2022/).
For these organizations, developers are their *customers*.
They routinely run surveys in order to tailor their products and services to their customer (e.g. developer) needs.

Academic institutions and national labs alike (or, frankly, anywhere else customized, software-enabled research is conducted) also have an interest in understanding developer needs.
This is true from a perspective of both developer productivity and workforce development and retention.

> ### Happy developers are productive developers?

Reviewer: ```suggestion
### Happy developers are productive developers
```
Should this be a question or a statement?

Developer needs can be grouped into two categories - the practical (e.g., tools, skills and resources) and the intangible (e.g., feelings of happiness, inspiration and inclusion).
The practical things are fairly straightforward to deal with in a survey.

But, we also want our developers to be happy, feel like they are working on meaningful projects, and have the resources they need.

> Working on a project that provides general support to other development teams (e.g. IT support for site access for remote workers), its easy to feel disconnected from the core research mission(s) at your institution.

Reviewer: ```suggestion
If an individual is working on a project that provides general support to other development teams (e.g. IT support for site access for remote workers) they might feel disconnected from the core research mission(s) at their institution.
If we don't regularly ask them about their experience, this disconnect may never be known.
```

Tunnel vision can develop that reflects only the immediate needs and priorities of your team and can even result in developing solutions for problems that don't exist!
It's OK, we've all been there.
This is why we need to survey developers about their happiness and needs.
When our developers are happy and have what they need to be successful, the work will be more productive and fun, and everyone wins.

### Why did LLNL do a survey this year?

I joined Lawrence Livermore National Lab (LLNL) right after the start of the pandemic.

> I knew then that I wanted to continue my basic research to understand open source software and developer needs -- research I had started at Stanford University with my first [software survey](https://stanford-rc.github.io/stanford-software-survey/).

Reviewer: ```suggestion
I knew then that I wanted to continue my basic research to understand open source software and developer needs -- work I had started at Stanford University with my first [software survey](https://stanford-rc.github.io/stanford-software-survey/).
```
It seems like  a far stretch to call it research!

> At LLNL, there had previously been an annual survey of just the flagship simulation code teams that resulted in a dense, internally distributed document of survey results in text and plots.

Reviewer: ```suggestion
At LLNL, there had previously been an annual survey of just application teams that resulted in a dense, internally distributed document of survey results in text and plots.
```

I saw a missed opportunity in that.

> I thought we might have more fun making an annual tradition of sending out a survey and collecting results (still internally) in a more interactive web interface giving the LLNL developer community yearly updates about what's going on in software development at LLNL.

Reviewer: ```suggestion
I thought we might have more fun making an annual tradition of sending out a survey and collecting results (still internally) in a more interactive web interface.
We could give the LLNL developer community yearly updates about what's going on in software development at LLNL.
```

> In addition, one code team in particular, the [RADIUSS](https://software.llnl.gov/radiuss/) team, served as a focal point for early work on the survey because the RADIUSS project involves a wide variety of interdependent LLNL software projects.

Reviewer: ```suggestion
In addition, one developer team in particular, the [RADIUSS](https://software.llnl.gov/radiuss/) team, served as a focal point for early work on the survey because the RADIUSS project supports a wide variety of interdependent LLNL software projects.
```
This isn't a code team.

While Likert scales can show general sentiments and trends, these open-ended responses often capture gems of developer feedback.
You really can't put people in a check box, nor can you ask them to express a need or sentiment with one.

Once the questions were designed, I used a Microsoft Form (a requirement given other resources I was marshaling) to construct the actual survey.
As an aside, if you have a choice, I find Microsoft Forms much harder to use than Google Forms.
We had a few initial rounds of testing and feedback starting with the RADIUSS team.
Many questions were refined, simplified, or entirely removed during this process.
Our perspective at this time was to do our best with available resources, and not worry about making it perfect.
We planned to learn from the experience and use what we learn for next time.
A few months before the end of 2022, the survey was ready to launch.

### Launch Away... to Results!

We sent out the survey in the last few months of 2022, making sure to give it fun branding.

> We sent requests for responses on many mailing lists and a number of other platforms where we could reach large numbers of LLNL developers (e.g. Town Hall channel, Teams, etc.).

Reviewer: ```suggestion
We sent requests for responses on many mailing lists and a number of other platforms where we could reach large numbers of LLNL developers (e.g. community chat channels, email lists, etc.).
```

We kept the survey open for about 2 months and sent out a few reminders during that time to encourage respondents to complete it.

> We don't have an accurate count of total number of developers we ultimately reached to solicit their input.

Reviewer: ```suggestion
```

> I suspect we reached a few hundred developers.

Reviewer: ```suggestion
```
This is not correct.

> I suspect the the majority of responses came from Livermore Computing which is the division of employees responsible for maintaining the leadership computing facilities at LLNL.

Reviewer: ```suggestion
```
This is also not correct!

> 

Reviewer: ```suggestion
```
Remove newline

We had over 100 responses, half of which were self-reported developers.
About a quarter were staff scientists (that also develop code) and the remainder of respondents included group leaders, managers, IT support, or other support staff.

> We had more participation than was seen in application surveys in previous years.

Reviewer: ```suggestion
We had more participation than was seen in application surveys in previous years, which to us was an indication of a successful effort.
```

Some of the results were expected - an overwhelmingly positive flow of gratitude for the people who lead or support tooling infrastructure and their quality of work, as well as gripes about developer environments and/or tooling.
I say "expected" because this is a manifestation of a classic dichotomy at a research organization like an academic institution or a national lab.
Anywhere shared high performance computing environments are used, there is a dual need to be conservative and reliable, and that is challenged by the constant reality and drive for change and new capabilities.

One of the biggest challenges in this process isn't necessarily knowing what are the biggest pain points, but figuring out how to address them with the available resources.
As an example, the RADIUSS team is small with each member only 10-50% time on the project and often only for a specific component of the whole ecosystem.
Even when we are aware of big challenges needing attention, it may take such a fragmented team a while to find the resources to fix it.
So we decided that we would evangelize key takeaways of the survey with teams and project leaders to help inform their priorities going forward.

> For example, one key takeaway is that developers really like the [VSCode](https://code.visualstudio.com) editor (not to be confused with the Visual Studio IDE).

Reviewer: ```suggestion
As an example, one strong piece of insight we got (that is public) is that developers really like VSCode.
```

> It was very easy to correlate this with [this announcement from GitLab](https://software.llnl.gov/radiuss/2023/02/08/first-developer-survey/)) and approach one of the leads of the team that maintains GitLab and ask if this could be provided.

Reviewer: ```suggestion
It was very easy for me to see this [this announcement from GitLab](https://devclass.com/2022/12/20/gitlab-goes-public-with-web-ide-beta-based-on-visual-studio-code/) and realize that getting this integration enabled for our private GitLab instances would be hugely valued.
```

> It was very easy to correlate this with [this announcement from GitLab]([a public facing](https://software.llnl.gov/radiuss/2023/02/08/first-developer-survey/)) and approach one of the leads of the team that maintains GitLab and ask if this could be provided.

Author: @vsoch...I didn't understand the above sentence.

After fixing the mechanics of the link, I also noticed it doesn't point to an announcement from GitLab. 

So, I may have fouled this up. But, it should be easy to fix if you can recall original.

> It was very easy to correlate this with [this announcement from GitLab]([a public facing](https://software.llnl.gov/radiuss/2023/02/08/first-developer-survey/)) and approach one of the leads of the team that maintains GitLab and ask if this could be provided.

Author: Fix link...

```suggestion
It was very easy to correlate this with [this announcement from GitLab](https://software.llnl.gov/radiuss/2023/02/08/first-developer-survey/)) and approach one of the leads of the team that maintains GitLab and ask if this could be provided.
```
